ios/CodePush/CodePushPackage.m

```diff
@@ -230,19 +244,36 @@ + (void)downloadPackage:(NSDictionary *)updatePackage
                                                                 return;
                                                             }
                                                         }
+
+                                                        if (isDiffUpdate) {
+                                                            CPLog(@"Applying diff update.");
+                                                        } else {
+                                                            CPLog(@"Applying full update.");
+                                                        }
                                                         
-                                                        if (isDiffUpdate && ![CodePushUpdateUtils verifyHashForDiffUpdate:newUpdateFolderPath
-                                                                                                             expectedHash:newUpdateHash
-                                                                                                                    error:&error]) {
-                                                            if (error) {
-                                                                failCallback(error);
+                                                        BOOL isSignatureVerificationEnabled = (publicKey != nil);
+                                                        if (isSignatureVerificationEnabled) {
+                                                            BOOL isSignatureValid = [CodePushUpdateUtils verifySignatureFor:newUpdateFolderPath
+                                                                                      withPublicKey:publicKey
+                                                                                              error:&error];
+                                                            if (!isSignatureValid) {
+                                                                [self handleFailedDataIntegrityCheck:failCallback error:&error];
```

Comment:
I would prefer to avoid of using `handleFailedDataIntegrityCheck` method at all and implement error handling like this:
```objective-c
    if (!error) {
        error = [CodePushErrorUtils errorWithMessage:@"Code Signing integrity check error..."];
    }

    CPLog(error);

    failCallback(*error);
    return;
``` 
This will allow us to have different error messages for code signing integrity check and for general integrity check determine in future where the issue occurs.

Reply:
@sergey-akhalkov nice proposition, I'll do it!